File: examples/minimart-go-nuxt/backend/internal/application/usecase/auth_usecase.go
// Package usecase is the application layer: it orchestrates repositories and
// domain rules and returns wire shapes. It knows nothing about gin, cookies or
// JWTs — a session is a transport concern and lives in infrastructure.
package usecase

import (
	"context"
	"errors"

	"golang.org/x/crypto/bcrypt"

	"minimart/internal/application/dto"
	"minimart/internal/domain/entity"
	"minimart/internal/infrastructure/repository"
)

// dummyHash is a valid bcrypt hash of a value nobody knows. It is compared
// against when the username does not exist, so that "no such user" costs the
// same time as "wrong password" and the endpoint does not leak which usernames
// are registered through its latency after refusing to leak it through its
// status code.
const dummyHash = "$2a$10$MwKxT13/lPxFMjvrkm0dL.QJuNll1zljDU.eOoRVvgrWF.kf/hyC2"

// AuthUsecase verifies credentials and reads the caller's own profile.
type AuthUsecase struct {
	users *repository.UserRepository
}

// NewAuthUsecase returns an AuthUsecase over users.
func NewAuthUsecase(users *repository.UserRepository) *AuthUsecase {
	return &AuthUsecase{users: users}
}

// Authenticate returns the user whose password matches, or
// ErrInvalidCredentials. An unknown username and a wrong password are the same
// answer, deliberately.
func (uc *AuthUsecase) Authenticate(ctx context.Context, username, password string) (*entity.User, error) {
	user, err := uc.users.FindByUsername(ctx, username)
	switch {
	case errors.Is(err, repository.ErrNotFound):
		_ = bcrypt.CompareHashAndPassword([]byte(dummyHash), []byte(password))
		return nil, entity.ErrInvalidCredentials
	case err != nil:
		return nil, err
	}
	if err := user.VerifyPassword(password); err != nil {
		return nil, err
	}
	return user, nil
}

// RoleOf returns the role stored on the caller's users row.
//
// It exists so that authorisation is answered from the database rather than
// from the token's role claim. The claim states what was true when the token
// was issued; a JWT cannot be withdrawn, so a user demoted an hour ago still
// carries one saying otherwise for the rest of the day.
//
// A subject with no row is ErrUnauthenticated, for the same reason Me gives:
// the session is what has become invalid.
func (uc *AuthUsecase) RoleOf(ctx context.Context, userID int) (string, error) {
	role, err := uc.users.FindRoleByID(ctx, userID)
	switch {
	case errors.Is(err, repository.ErrNotFound):
		return "", entity.ErrUnauthenticated
	case err != nil:
		return "", err
	}
	return role, nil
}

// Me returns the profile of the authenticated caller.
//
// A token whose subject no longer exists is UNAUTHENTICATED, not 404: the
// session is the thing that is invalid, and the caller's remedy is to log in
// again rather than to look for a different user.
func (uc *AuthUsecase) Me(ctx context.Context, userID int) (*dto.MeResponse, error) {
	user, err := uc.users.FindByID(ctx, userID)
	switch {
	case errors.Is(err, repository.ErrNotFound):
		return nil, entity.ErrUnauthenticated
	case err != nil:
		return nil, err
	}
	return &dto.MeResponse{
		Username: user.Username,
		Role:     user.Role,
		Points:   user.Points,
	}, nil
}

File: examples/minimart-java-next/backend/src/main/java/minimart/application/usecase/AuthUsecase.java
package minimart.application.usecase;

import java.util.Optional;
import minimart.application.dto.MeResponse;
import minimart.domain.DomainException;
import minimart.domain.ErrorCode;
import minimart.domain.User;
import minimart.infrastructure.repository.UserRepository;

/** Verifies credentials and reads the caller's own profile. */
public class AuthUsecase {

  /**
   * A valid bcrypt hash of a value nobody knows. It is compared against when the username does not
   * exist, so that "no such user" costs the same time as "wrong password" and the endpoint does not
   * leak which usernames are registered through its latency after refusing to leak it through its
   * status code.
   */
  static final String DUMMY_HASH = "$2a$10$MwKxT13/lPxFMjvrkm0dL.QJuNll1zljDU.eOoRVvgrWF.kf/hyC2";

  /** The stand-in whose password check is run, and whose result is ignored, for unknown names. */
  private static final User NOBODY = new User(0, "", DUMMY_HASH, User.ROLE_GUEST, 0);

  private final UserRepository users;

  /** A use case over {@code users}. */
  public AuthUsecase(UserRepository users) {
    this.users = users;
  }

  /**
   * Returns the user whose password matches.
   *
   * @throws DomainException INVALID_CREDENTIALS for an unknown username and for a wrong password
   *     alike, deliberately
   */
  public User authenticate(String username, String password) {
    Optional<User> found = users.findByUsername(username);
    if (found.isEmpty()) {
      // Spend the bcrypt cost anyway; see DUMMY_HASH. The result is meaningless and discarded.
      NOBODY.verifyPassword(password);
      throw new DomainException(ErrorCode.INVALID_CREDENTIALS);
    }
    User user = found.get();
    if (!user.verifyPassword(password)) {
      throw new DomainException(ErrorCode.INVALID_CREDENTIALS);
    }
    return user;
  }

  /**
   * Returns the role stored on the caller's users row.
   *
   * <p>It exists so that authorisation is answered from the database rather than from the token's
   * role claim. The claim states what was true when the token was issued; a JWT cannot be
   * withdrawn, so a user demoted an hour ago still carries one saying otherwise for the rest of the
   * day.
   *
   * @throws DomainException UNAUTHENTICATED when the subject has no row, for the same reason {@link
   *     #me} gives: the session is what has become invalid
   */
  public String roleOf(long userId) {
    return users
        .findRoleById(userId)
        .orElseThrow(() -> new DomainException(ErrorCode.UNAUTHENTICATED));
  }

  /**
   * Returns the profile of the authenticated caller.
   *
   * @throws DomainException UNAUTHENTICATED, not 404, for a subject that no longer exists: the
   *     session is the thing that is invalid, and the caller's remedy is to log in again rather
   *     than to look for a different user
   */
  public MeResponse me(long userId) {
    User user =
        users.findById(userId).orElseThrow(() -> new DomainException(ErrorCode.UNAUTHENTICATED));
    return new MeResponse(user.username(), user.role(), user.points());
  }
}
